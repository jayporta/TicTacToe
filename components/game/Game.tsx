import React, { useState } from 'react';
import { TouchableOpacity, Text, View } from 'react-native';
import { Players } from '../../types';
import styles from './gameStyles';
import calculateWinner from './calculateWinner';

const squaresDefault = Array(9).fill(null) as string[];

const Game = ({
  players,
  resetGame,
}: {
  players: Players;
  resetGame: () => void;
}) => {
  const [squares, setSquares] = useState([...squaresDefault]);
  const [activePlayer, setActivePlayer] = useState('X');

  const handlePress = (index: number) => {
    if (squares[index]) {
      return;
    }
    const squareSpliced = squares.slice();
    squareSpliced[index] = activePlayer;
    setSquares(squareSpliced);
    setActivePlayer(activePlayer === 'X' ? 'O' : 'X');
  };

  const winner = calculateWinner(squares);

  if (winner) {
    const handlePressPlayAgain = () => {
      setSquares(squaresDefault);
      resetGame();
    };

    return (
      <View style={styles.container}>
        <Text style={styles.winner}>{`${
          players[winner as keyof Players]
        } has won!`}</Text>
        <TouchableOpacity
          style={styles.playAgain}
          onPress={handlePressPlayAgain}
          accessibilityRole="button"
          accessibilityHint="Tap to start over"
        >
          <Text style={styles.playAgainText}>Play Again</Text>
        </TouchableOpacity>
      </View>
    );
  }

  return (
    <View style={styles.container}>
      {
        <Text style={styles.title}>{`${
          players[activePlayer as keyof Players]
        }'s turn (${activePlayer})`}</Text>
      }
      <View style={styles.board}>
        {squares.map((mark, i) => (
          <TouchableOpacity
            key={i}
            style={styles.square}
            onPress={() => handlePress(i)}
            accessibilityHint={`Tap to place an ${mark}`}
            testID={`square${i}`}
          >
            <Text style={styles.squareText}>{mark || ''}</Text>
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );
};

export default Game;
