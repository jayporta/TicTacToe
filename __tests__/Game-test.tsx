import React from 'react';
import 'react-native';
import { fireEvent, render } from '@testing-library/react-native';
import Game from '../components/game/Game';
import renderer from 'react-test-renderer';

it('Game matches snapshot', () => {
  const resetMock = jest.fn();
  const tree = renderer
    .create(<Game players={{ X: '', O: '' }} resetGame={resetMock} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('displays winner by tic tac toe rules', () => {
  const { getByTestId, getByText } = render(
    <Game players={{ X: 'Jay', O: 'Kyla' }} resetGame={jest.fn} />
  );
  fireEvent.press(getByTestId('square0')); // X
  fireEvent.press(getByTestId('square3')); // O
  fireEvent.press(getByTestId('square1')); // X
  fireEvent.press(getByTestId('square6')); // O
  fireEvent.press(getByTestId('square2')); // X
  expect(getByText('Jay has won!')).toBeTruthy();
});
