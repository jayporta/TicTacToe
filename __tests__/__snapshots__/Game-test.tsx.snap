// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game matches snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
        "marginBottom": 50,
      }
    }
  >
    's turn (X)
  </Text>
  <View
    style={
      Object {
        "borderStyle": "solid",
        "borderWidth": 1,
        "display": "flex",
        "flexDirection": "row",
        "flexWrap": "wrap",
        "width": 302,
      }
    }
  >
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square0"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square1"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square2"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square3"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square4"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square5"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square6"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square7"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
    <View
      accessibilityHint="Tap to place an null"
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderStyle": "solid",
          "borderWidth": 1,
          "display": "flex",
          "height": 100,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
      testID="square8"
    >
      <Text
        style={
          Object {
            "fontSize": 30,
          }
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
